microcosm_eventsource/immutable_event_factory.py

```diff
@@ -0,0 +1,81 @@
+"""
+Factory for immutable events.
+
+In extension of EventFactory, this class provides a native support of event updating container objects based
+on event types
+
+"""
+from collections import defaultdict
+from typing import Dict
+
+from microcosm_eventsource.factory import EventFactory
+
+
+container_common_mutator_registry: Dict[str, list] = defaultdict(list)
+container_event_specific_mutator_registry: Dict[str, list] = defaultdict(list)
+
+
+def register_container_mutator_common(event_type):
+    def decorator(func):
+        container_common_mutator_registry[event_type.__name__].append(func.__name__)
+        return func
+
+    return decorator
+
+
+def register_container_mutator_by_event_type(event_instance_type):
+    def decorator(func):
+        container_event_specific_mutator_registry[event_instance_type.name].append(func.__name__)
+        return func
+
+    return decorator
+
+
+def get_specific_handler_names_for_event_type(event):
+    return container_event_specific_mutator_registry[event.event_type.name]
+
+
+def get_common_handler_names_for_event(event):
+    event_type_name = f"{event.__class__.__container__.__name__}Event"
+    return container_common_mutator_registry[event_type_name]
+
+
+class ImmutableEventFactory(EventFactory):
```

Comment:
I'm confused by the meaning of `immutable` here. Isn't this in fact an event factor which should do mutations (to a container object)? Maybe a name like "AggregatingEventFactory"? which aggregates event state on an events `container`?

Reply:
How about ContainerMutatorEventFactory? I agree ImmutableEventFactory does not convey the purpose of the class.